### PR TITLE
1277 malawi patient profile

### LINKED ISSUE #1277: MWCore Patient Profile
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

- [ ] Show the following patient details
    -ART Number
    -Full name
    -Age
    -Gender

- [ ] Be able to update existing patient details
- [ ] Add menu option for 'Client Clinic Visit' if the patient is either a newly diagnosed client or a client on ART
- [ ] Add menu option for 'Exposed Infant Clinic Visit' if the patient is an exposed infant
- [ ] Add menu option for 'Guardian Visit' if the patient is a newly diagnosed client or a client on ART
- [ ] Add a menu option for 'Sample Collection & Lab Results' for all patients
- [ ] Add a menu option for 'Clinic History' for all patients

**Additional context**
Please view this ppt with [desired workflow](https://docs.google.com/presentation/d/1cObLZgSWztlBY7t8T-KiOn0_VYnLjFKj/edit#slide=id.p1)

### PR DESCRIPTION
**Pull request by AbdulWahabMemon**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #1277 

**Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/patient/profile/PatientProfileViewModel.kt [hunk 1]
```diff
@@ -74,11 +75,45 @@ constructor(
...
   }
 
+  fun refreshOverFlowMenu(healthModule: HealthModule, patientProfile: ProfileData) {
+    if (healthModule == HealthModule.HIV ||
+        healthModule == HealthModule.TRACING ||
+        healthModule == HealthModule.APPOINTMENT
+    ) {
+      val patientMetaFilterType = (patientProfile as ProfileData.DefaultProfileData).filterType
+      when {
+        patientMetaFilterType.equals("exposed-infant", true) ||
```
Comment (AbdulWahabMemon) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/patient/profile/PatientProfileViewModel.kt:91]: exposed-infant, child-contact, other types to filtering patient here can be check with enum constants,

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/PatientType.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  NEWLY_DIAGNOSED_CLIENT,
+  CLIENT_ALREADY_ON_ART,
+  EXPOSED_INFANT,
+  COMMUNITY_POSITIVE,
+  CHILD_CONTACT,
+  SEXUAL_CONTACT,
+  DEFAULT
+}
+
+// enum class PatientType(value: String, display: String = "") {
```
Comment (AbdulWahabMemon) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/PatientType.kt:29]: will remove in next commit

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/register/dao/DefaultPatientRegisterDao.kt [hunk 3]
```diff
@@ -90,7 +91,7 @@ constructor(
         identifier = patient.identifierFirstRep.value,
         address = patient.extractAge(),
         gender = patient.gender,
-        birthdate = patient.birthDate,
+        birthdate = patient.birthDate ?: Date(),
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/register/dao/DefaultPatientRegisterDao.kt:94]: Can we make this nullable instead? Defaulting to today's date would be incorrect.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/PatientType.kt (hunk 2 above)
Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/PatientType.kt:29]: We will need to configure this.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt [hunk 4]
```diff
@@ -211,3 +212,14 @@ fun Patient.extractOfficialIdentifier(): String? =
   if (this.hasIdentifier())
     this.identifier.firstOrNull { it.use == Identifier.IdentifierUse.OFFICIAL }?.value
   else null
+
+fun Patient.extractTypeViaDTreeMeta(): PatientType {
+  return try {
+    val tagList = this.meta.tag.filter { it.system.equals("https://d-tree.org", true) }
+
+    PatientType.valueOf(tagList[0].code?.uppercase(Locale.getDefault())?.replace("-", "_") ?: "")
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt:220]: This code is error prone. Empty string cannot be converted to `PatientType`

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt [hunk 5]
```diff
@@ -211,3 +212,14 @@ fun Patient.extractOfficialIdentifier(): String? =
...
+  } catch (e: Exception) {
+    e.printStackTrace()
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt:222]: Replace with `Timber`

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/register/dao/DefaultPatientRegisterDao.kt (hunk 3 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/register/dao/DefaultPatientRegisterDao.kt:94]: its reverted, but couldn't push yesterday, doing shortly, 
I added to bypass crash as its been crashing on most of patient details, (client list to profile screen)

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt (hunk 5 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt]: fixed

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt (hunk 4 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt:221]: fixed

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/PatientType.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+// }
+
+
+
+//                code: "newly-diagnosed-client", display: "Newly Diagnosed Client"
+//                code: "client-already-on-art", display: "Client Already On ART"
+//                code: "exposed-infant", display: "Exposed Infant"
+//                code: "community-positive", display: "Community Positive"
+//                code: "child-contact", display: "Child Contact"
+//                code: "sexual-contact", display: "Sexual Contact"
```
[the remark this answers is not in the document]
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/PatientType.kt]: removed

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt [hunk 7]
```diff
@@ -211,3 +213,14 @@ fun Patient.extractOfficialIdentifier(): String? =
   if (this.hasIdentifier())
     this.identifier.firstOrNull { it.use == Identifier.IdentifierUse.OFFICIAL }?.value
   else null
+
+fun Patient.extractTypeViaDTreeMeta(): PatientType {
+  return try {
+    val tagList = this.meta.tag.filter { it.system.equals("https://d-tree.org", true) }
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt:219]: I think we should not hard code this `"https://d-tree.org"`

Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt:219]: Can we obtain it from application configuration? logged in user attribute or something?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt [hunk 8]
```diff
@@ -211,3 +213,14 @@ fun Patient.extractOfficialIdentifier(): String? =
   if (this.hasIdentifier())
     this.identifier.firstOrNull { it.use == Identifier.IdentifierUse.OFFICIAL }?.value
   else null
+
+fun Patient.extractTypeViaDTreeMeta(): PatientType {
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt:217]: This should be an extension function and should be reusable.

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt (hunk 8 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt:217]: its already as extension, though need to generalize and remove "Dtree"

**Comment by AbdulWahabMemon** [comment]
this PR is also updated with below issues mentioned from Khumboo(D-Tree team) in email thread
cc: @Gental-Giant 

- There is no "Home Tracing" option in the navigation drawer.
- The label "Clients List" should be "Client List".
- The Client List icon should be the patient's icon that was there before and not the home icon that should be used for "Home Tracing".
- The list of patients in the Client List also shows in the tracing as well as appointments pages.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt (hunk 7 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/PatientExtension.kt]: fixed

**Comment by dubdabasoduba** [comment]
> @AbdulWahabMemon Do you have screenshots for this? We need to review what has changed from the default profile we have

@AbdulWahabMemon ??

**Comment by AbdulWahabMemon** [comment]
> > > @AbdulWahabMemon Do you have screenshots for this? We need to review what has changed from the default profile we have
> > 
> > 
> > @AbdulWahabMemon ??
> 
> there is no such profile display changes as compare with default, though profile detail menu option can be varies depending on Profile Patient Type, I will be running the app again and will share the snap for it.

**Comment by AbdulWahabMemon** [comment]
@dubdabasoduba , 

Screen Shot 2022-06-02 at 3 41 55 PM
Screen Shot 2022-06-02 at 3 42 23 PM
Screen Shot 2022-06-02 at 3 50 51 PM

**Review by maimoonak — DISMISSED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt [hunk 9]
```diff
@@ -82,11 +82,14 @@ sealed class ProfileData(open val logicalId: String, open val name: String) {
...
     val birthdate: Date,
-    val age: String,
+    val age: String = birthdate.toAgeDisplay(),
     val gender: Enumerations.AdministrativeGender,
     val address: String,
     val phoneContacts: List<ContactPoint> = listOf(),
-    val chwAssigned: Reference
+    val chwAssigned: Reference,
+    val patientType: PatientType,
+    val services: List<CarePlan> = listOf(),
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt:91]: it makes sense that we'd need this, how are we handling this now though?

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/patient/profile/PatientProfileViewModel.kt [hunk 10]
```diff
@@ -74,11 +76,44 @@ constructor(
               profileViewDataMapper.transformInputToOutputModel(it) as
                 ProfileViewData.PatientProfileViewData
             // TODO only display some overflow menu items when certain conditions are met
+            refreshOverFlowMenu(healthModule = healthModule, patientProfile = it)
           }
       }
     }
   }
 
+  fun refreshOverFlowMenu(healthModule: HealthModule, patientProfile: ProfileData) {
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/patient/profile/PatientProfileViewModel.kt:85]: this is definitely too custom, we need to figure out a generic way to manage this

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/PatientType.kt [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.engine.domain.model
+
+enum class PatientType {
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/PatientType.kt:19]: this looks too domain specific

**Review by pld — CHANGES_REQUESTED** [review]

**Comment by AbdulWahabMemon** [comment]
> * Newly Diagnosed Client: Client Clinic Visit, Guardian Visit, Viral Load Results, Edit Profile, Guardian Management, Clinic History
> * Client Already On ART: Client Clinic Visit, Guardian Visit, Viral Load Results, Edit Profile, Guardian Management, Clinic History
> * Exposed Infant: Exposed Infant Clinic Visit, HIV Test and Results, Edit Profile, Guardian Management, Clinic History
> * Child Contact: HIV Test and Next Appointment, Edit Profile, Guardian Management, Clinic History
> * Sexual Contact: HIV Test and Next Appointment, Edit Profile, Clinic History
> * Community Positive: HIV Test and Next Appointment, Edit Profile, Guardian Management, Clinic History

Screen Shot 2022-06-02 at 7 16 52 PM
Screen Shot 2022-06-02 at 7 17 09 PM

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt (hunk 9 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt:91]: Hi @pld , its is not in use for now, as it was requested and discussed before that patient profile detail would be require with field of assigned-chw, this property is added there, further @KhumboLihonga can better comment how they plan to use chw in profile data.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt [hunk 12]
```diff
@@ -82,11 +82,14 @@ sealed class ProfileData(open val logicalId: String, open val name: String) {
     override val name: String,
     val identifier: String? = null,
     val birthdate: Date,
-    val age: String,
+    val age: String = birthdate.toAgeDisplay(),
     val gender: Enumerations.AdministrativeGender,
     val address: String,
     val phoneContacts: List<ContactPoint> = listOf(),
-    val chwAssigned: Reference
+    val chwAssigned: Reference,
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt:89]: specifically on chwAssigned, can we manage this through linking a CareTeam to the Patient and putting the Practitioner (CHW) in that CareTeam?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt (hunk 9 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt:91]: Care plans are supposed to be use for different clinic visit of a patient, 
@Gental-Giant can comment on this

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt (hunk 12 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt:89]: need some guidance or sample to map that with careteam

**Comment by KhumboLihonga** [comment]
Here are my comments:
- unique id (identifier.value attribute in Patient resource not the id) to be visible only if it exists
- unique id to be shown alongside  'sex, age, DOB' and not underneath it
- 'DOB' to be removed
- even though the Patient has the gender attribute assigned, 'Sex' is blank
- 'Sex' to be re-labelled to 'Gender'

**Comment by dubdabasoduba** [comment]
> Here are my comments:
> 
> * ART Number/HCC Number/HTS Number (identifier.value attribute in Patient resource not the id) to be visible only if it exists
> * ART Number/HCC Number/HTS Number to be shown alongside  'Sex, Age, DOB' and not underneath it
> * 'DOB' to be removed
> * even though the Patient has the gender attribute assigned, 'Sex' is blank
> * 'Sex' to be re-labelled to 'Gender'

@AbdulWahabMemon @KhumboLihonga 

> ART Number/HCC Number/HTS Number to be shown alongside 'Sex, Age, DOB' and not underneath it

Attached is our individual profile design. We will like to maintain the design based on user feedback we've gathered from previous OpenSRP implementations. If you notice the ID come just below the Patient's name. 
Individual details - tasks (1)

Do you mind elaborating more on why you think having it in the same position as the Age, Gender and DOB makes more sense? 

> 'Sex' to be re-labelled to 'Gender'

@AbdulWahabMemon Please copy the string from Quest and update the `strings.xml` on the MWCore flavour. This will be sufficient as we define the configurability pieces on the profile. 

> ART Number/HCC Number/HTS Number (identifier. value attribute in Patient resource, not the id) to be visible only if it exists

@AbdulWahabMemon I think currently the profile shows the identifier that's on position 0 in the identifiers array. This identifier should also be the `official` idenfier.  Please update your StructureMap to reflect the same. 

> even though the Patient has the gender attribute assigned, 'Sex' is blank

@AbdulWahabMemon This sounds like a bug. Please look into it. The FHIRCore non-project specific profile seems to work well.

**Comment by AbdulWahabMemon** [comment]
@dubdabasoduba , Gender is showing and that was fix in yesterday comment,
Sex is re-labelled to Gender using flavor, that is also pushed, 
updated screenshot attached. 
ProfileView attribute showDOBInProfile is set false for HivRegisterDoa which comes in use HIV HealthModule, use only in mwcore config

Screen Shot 2022-06-08 at 1 18 40 PM

**Comment by AbdulWahabMemon** [comment]
> > Here are my comments:
> > 
> > * ART Number/HCC Number/HTS Number (identifier.value attribute in Patient resource not the id) to be visible only if it exists
> > * ART Number/HCC Number/HTS Number to be shown alongside  'Sex, Age, DOB' and not underneath it
> > * 'DOB' to be removed
> > * even though the Patient has the gender attribute assigned, 'Sex' is blank
> > * 'Sex' to be re-labelled to 'Gender'
> 
> @AbdulWahabMemon @KhumboLihonga
> 
> > ART Number/HCC Number/HTS Number to be shown alongside 'Sex, Age, DOB' and not underneath it
> 
> Attached is our individual profile design. We will like to maintain the design based on user feedback we've gathered from previous OpenSRP implementations. If you notice the ID come just below the Patient's name. Individual details - tasks (1)
> 
> Do you mind elaborating more on why you think having it in the same position as the Age, Gender and DOB makes more sense?
> 
> > 'Sex' to be re-labelled to 'Gender'
> 
> @AbdulWahabMemon Please copy the string from Quest and update the `strings.xml` on the MWCore flavour. This will be sufficient as we define the configurability pieces on the profile.
> 
> > ART Number/HCC Number/HTS Number (identifier. value attribute in Patient resource, not the id) to be visible only if it exists
> 
> @AbdulWahabMemon I think currently the profile shows the identifier that's on position 0 in the identifiers array. This identifier should also be the `official` idenfier. Please update your StructureMap to reflect the same.
> 
> > even though the Patient has the gender attribute assigned, 'Sex' is blank
> 
> @AbdulWahabMemon This sounds like a bug. Please look into it. The FHIRCore non-project specific profile seems to work well.

Sex relabelled to Gender, also pushed. @dubdabasoduba

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]
@pld Based on this screenshot. We might have to block the merging of this PR to the main branch. It changes the overall product profile design. 

172568050-f5ea66be-aaa9-438b-8ba3-b8228502274a

@Gental-Giant Please set up a call for us to discuss this.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt (hunk 12 above)
Comment (dubdabasoduba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ProfileData.kt]: @Gental-Giant @pld 

I had a call with @AbdulWahabMemon here are the notes. 

**Notes**
**Questions to answer**
- Are CHWs allowed to assign other CHWs patients. 

**Registration workflow (Assigning to Care teams)**
**FHIR Web**
- Create the CHW (Practitioner)
- Create the patient category groups
- Create the CHW care teams
    - Assign the groups above to the correct care team

**FHIR Core**
**Registration**
- During patient registration, add the patient to the correct patient category group
- On CHW assigned selection, do we show the CHWs on the specific care team or do we show all the CHWs? 

**Profile**
- We can add a section called an assignment. 
    - This should show the care team that is assigned to the group to which the patient is assigned. 



**Registration workflow (Assigning to Practitioner)**
**FHIR Web**
- Create the CHW (Practitioner)

**FHIR Core** 
**Registration**
- During patient registration, assign the patient to the correct practitioner 
- On CHW assigned selection (Not dynamic currently)
    - Show a list of all practitioners in the app. 
    - This list should be synced locally from the server. 

**Profile**
- We can add a section called an assignment. 
   - This should show the practitioner the patient is assigned to. 


@pld Based on the conversation above we might need to add a section on the profile to show the assignments. It should be configurable to show individual assignments or team assignments. 

@Gental-Giant 
- Please check in with the client on the Question above. 
- Also check in with @rowo @HenryRae about the best place to have the assignments shown. 


@pld I think we can leave the implementation as is till we figure out the correct workflow.

**Comment by ellykits** [comment]
Tagging this with `DNM` label until the discussion is settled.

**Comment by KhumboLihonga** [comment]
@ellykits apologies for the late response, I was out of office. We don't have to position the unique identifier alongside age and gender, I just thought it looked better. Positioning the ID underneath the name is fine. However, is it okay to re-label the ID as ART/HCC/HTS Number instead?

**Comment by pld** [comment]
> @ellykits apologies for the late response, I was out of office. We don't have to position the unique identifier alongside age and gender, I just thought it looked better. Positioning the ID underneath the name is fine. However, is it okay to re-label the ID as ART/HCC/HTS Number instead?

As long as it can handle the horizontal space that won't be a problem, the text will be configurable

**Review by pld — APPROVED** [review]

**Comment by AbdulWahabMemon** [comment]
Screen Shot 2022-06-27 at 3 31 56 PM
Screen Shot 2022-06-27 at 3 32 19 PM
Screen Shot 2022-06-27 at 3 38 03 PM


as of last updates pushed on Friday, which also covering @KhumboLihonga's points below, 

- I added a community positive patient named "Alexander Mokowena" but on viewing his profile, the id is 'ART Number'. Community positive patients will have an ID as 'HTS Number'. 
- On the patient profile, we don't need the DOB, you can remove that.


cc: @Gental-Giant , @f-odhiambo

**Review by pld — APPROVED** [review]

**Pull request merged**